py4ai/core/data/layer/v0/__init__.py

```diff
@@ -0,0 +1,194 @@
+        """
+        ...
+
+    def map(self, f: Callable[[T], V], *args: Any, **kwargs: Any) -> Iterator[V]:
+        """
+        Apply a tranformation to each object and return an iterator.
+
+        :param f: function
+        :param args: other arguments passed to the retrieve method
+        :param kwargs: other keyworded arguments passed to the retrieve method
+        :yield: iterator over transformed objects
+        """
+        for obj in self.retrieve(*args, **kwargs):  # type: T
+            yield f(obj)
+
+    def foreach(self, f: Callable[[T], None], *args, **kwargs) -> None:
+        """
+        Apply a function to each object.
+
+        :param f: function
+        :param args: other arguments passed to the retrieve method
+        :param kwargs: other keyworded arguments passed to the retrieve method
+        """
+        for obj in self.retrieve(*args, **kwargs):  # type: T
+            f(obj)
+
+    def retrieveGenerator(self, condition: Any, sort_by: Any) -> IterGenerator[T]:
+        """
+        Wrap the retrieve method to return an iterator generator.
+
+        :param condition: filter condition
+        :param sort_by: sorting options
+        :return: iterator generator
+        """
+
+        def _iterator():
+            return self.retrieve(condition=condition, sort_by=sort_by)
+
+        return IterGenerator(_iterator)
+
+
+class TableABC(ABC):
+    """Abstract class for tables."""
+
+    @abstractmethod
+    def to_df(self, query: str) -> pd.DataFrame:
+        """
+        Transform the underling data into a pandas dataframe.
+
+        :param query: query
+        :return: pandas dataframe
+        """
+        ...
+
+    @abstractmethod
+    def write(self, df: pd.DataFrame) -> None:
+        """
+        Write the pandas dataframe in the persistance layer.
+
+        :param df: pandas dataframe
+        """
+        ...
+
+
+class DatabaseABC(ABC):
+    """Abstract class for databases."""
+
+    @abstractmethod
+    def table(self, table_name: str) -> Optional[TableABC]:
+        """
+        Return the table with a given name.
+
+        :param table_name: name of the table
+        :return: table or None if table not found
+        """
+        ...
+
+
+class Writer(ABC):
+    """Abstract class to write Tables."""
+
+    @property
+    @abstractmethod
+    def table(self) -> TableABC:
+        """Return table object."""
+        ...
+
+    @abstractmethod
+    def push(self, df: pd.DataFrame) -> None:
+        """
+        Push a pandas dataframe.
+
+        :param df: dataframe
+        """
+        ...
+
+
+class EmptyDatabase(DatabaseABC):
+    """Class for empty Databases."""
+
+    def table(self, table_name: str) -> None:
+        """
+        Raise an exception, since the database is empty.
+
+        :param table_name: name of the table
+        :raises NoTableException: no table is present in an empty database
+        """
+        raise NoTableException(f"No table found with name {table_name}")
```

Comment:
I think that the whole v0 module should be dropped.

Reply:
Done. However I'd like to keep a commit about this. I believe also the v0 is a relevant piece of code that it would be good to have somewhere in our repo at the same time that the v1, to allow for comparisons eventually